### ISSUE TITLE
concatenate zarr datasets

### ISSUE THREAD

**Issue by ivirshup** [fsspec/kerchunk#216]
I would like to virtually concatenate a few zarr arrays along an axis. They are specifically not xarray datasets, and do not have coordinate arrays. 

Is this in scope for `kerchunk`s API?

**Comment by martindurant** [comment]
Yes, this is certainly something that kerchunk should be able to do. It can already "scan" zarr datasets as input (which is super simple compared to other file types) and then you can pass these to MultiZarrToZarr. Since you don't have coordinates in the inputs, you will need to figure out some other way to know where in the output each input is - and kerchunk will generate xarray-like coordinate arrays anyway. The `coo_map` argument is the one you would use to define this.

**Comment by ivirshup** [comment]
Is it expected to work for individual arrays? It seems to throw a `ContainsArrayError`. So it wants a group, but the groups I have will have multiple arrays which will need different treatment.

I would also want to avoid the xarray coordinate arrays in the output.

**Comment by martindurant** [comment]
> Is it expected to work for individual arrays

I would not be surprised if the code assumes groups. We ought to have a way to select amongst the members of a group for "valid" input on each run.

> avoid the xarray coordinate arrays in the output

We allow general postprocessing of the references after the combine process and before write, and dropping variables is so common, there is even a helper function. Also, editing the generated JSON by hand is simple enough.

**Comment by ivirshup** [comment]
What about a lower level api? Something like:

```python
def concatenate(
        arrays: list[zarr.Array],
        axis: int = 0
    ) -> dict:
    """Virtually concatenate arrays along axis, returning a kerchunk references dict"""
```

I would like to use this programmatically, but without needing to know exactly what xarray is expecting.

---------

Is there somewhere I can find more examples of `MultiZarrToZarr`? I'm not sure I'm understanding how any of the coordinate selection is supposed to work for this kind of case.

**Comment by martindurant** [comment]
I would be happy to see a simplified combine/concat like you suggest. Except, the inputs are not zarr arrays but reference sets (JSON or dict); or would you really want to special-case for zarr alone? Even for the latter, you would want to pass the URL/store rather than the array, since you need access to the filesystem object.

> Is there somewhere I can find more examples of MultiZarrToZarr

There are not yet enough, and they are indeed very netCDF layout oriented (cc @peterm790 ).

**Comment by ivirshup** [comment]
> Except, the inputs are not zarr arrays but reference sets (JSON or dict); 

Doesn't the set of references for any array map to a zarr array?

If they can, this wouldn't be special casing for zarr alone, but we would get the benefit of having an object that exposes things like `chunks`, `shape`, `dtype`, etc.

> Even for the latter, you would want to pass the URL/store rather than the array, since you need access to the filesystem object.

Shouldn't the array cary this information in `.store` and `.path`?

**Comment by martindurant** [comment]
> Shouldn't the array carry this information in .store and .path?

Yes, you can go that way around, but the store could be a mapper or an FSStore or consolidated...

**Comment by ivirshup** [comment]
So, maybe a `zarr.Array` with a restricted set of valid store types?

**Comment by ivirshup** [comment]
> I would be happy to see a simplified combine/concat like you suggest. Except, the inputs are not zarr arrays but reference sets (JSON or dict);

Practically, how would you see this happening? Would you or someone already working on the project be interested in implementing this, or is this more "PRs welcome"?

I could try, but can't promise any results, since I'm still mainly hitting errors with `MultiZarrToZarr` and have no experience with fsspec 😅

**Comment by martindurant** [comment]
I may attempt it today, if it looks like it may be as simple as I assume.
